rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -0,0 +1,490 @@
+
+* We collect system information from users on a case by case basis. This
+  automates some of that information gathering.
+* Diagnostics information will help us discover unusual user systems that we
+  might otherwise struggle to reproduce.
+
+#### Downsides
+
+* The benefits of a `diagnostic` sub-command are not immediate, are focused on
+  the after-release side of regression and some users will not wish to share
+  its output with us.
+* A `diagnostic` sub-command is not a solution in itself but must be paired
+  with other approaches.
+
+### Run test-harness Nightly
+
+In this alternative we take the existing test-harness code base and adjust it to
+run nightly (likely UTC 00:00 for convenience), building from the current head
+of master branch. We will need to actually build a nightly Vector release for
+use by test-harness, but this seems straightforward to achieve.
+
+Addresses:
+
+* Regression from Increased Instrumentation
+* musl libc Release Builds
+* Channel Implementation Regresses Topology Throughput
+
+Does not address:
+
+* `metrics` Crate Upgrade Regresses Benchmarks
+* Lua Transform Leaked Memory
+
+#### Upsides
+
+* By adjusting the test-harness to run nightly we reduce the total number of
+  commits that can potentially introduce regressions, naturally simplifying the
+  debugging process.
+* We need to make very few changes to the test-harness to achieve this, other
+  than rigging up a method for periodic execution.
+
+#### Downside
+
+* Other than reducing the number of commits that must be examined, this
+  approach retains all the defects of the existing test-harness.
+
+### Rethink test-harness
+
+Consider that the test-harness currently serves three purposes: correctness
+testing, performance regression testing and product comparison testing. The
+first two have value with every commit, the third is valuable when we update
+release documentation as a guide for new users and for keeping abreast of the
+progress of our competition. In this alternative we:
+
+* Make a logical, if not structural, split in the project between the
+  different methods of testing.
+* Exploit this split to run correctness and performance tests in a "nightly"
+  fashion, "comparison" tests for pre-releases or otherwise.
+* Exploit this split to run only Vector in performance testing, ensuring that
+  these tests are more straightforward to write and allowing us to write more
+  of them as the cost to introduce each goes down.
+* Run multiple instances of the same test and for significantly longer
+  duration, applying statistical controls after the fact to get cleaner
+  signal.
+* Significantly expand the data collection we do from the subject host
+  machine. We want to collect the same kinds of information that dstat does
+  and then more, ideally with an aim to understanding the internal mechanism
+  of Vector and how it plays against system resource constraints, in the [USE
+  method](http://www.brendangregg.com/usemethod.html) sense.
```

Comment:
It'd also be helpful to have more targeted and/or precise data from these runs. Right now they all return a wide variety of system-level data, not all of which is relevant to the current test, and which doesn't make it very clear which numbers are the ones we should focus on (i.e. throughput). That doesn't mean we shouldn't collect all that data and more, but we can simultaneously work to reduce the burden on the human interpreting the results.

Reply:
I agree. As described [here](https://github.com/timberio/vector/pull/6549#issuecomment-786919724) I intend for the nightly performance tests to track "big" numbers, something that's close to user concern and simple for us to see up/down whether there's been a regression -- future work can do linear regression and show trend lines and other fancy work -- and collect indicative side-channel information, probably into Datadog dashboards. 

The workflow I imagine us following: 

* Nightly reports a big swing in some perf test. 
* Interrogate reported side-channel numbers and see if there's some obvious aberration. 
  * If yes, can we make a smaller perf test that demonstrates the same issue? (This actually indicates a need to run ad-hoc experiments in this rig in a way I didn't fully appreciate.) 
  * If no, what resources would this test be constrained on and what do our detail numbers say about those resources? Were their any relevant swings in the criterion benchmarks? What do the flamegraphs of vector say? What about vector's self-telemetry?
* Form a hypothesis about possible vector issues, make focused experiments to invalidate them. Loop back to the second point. (Also another indication for the need to run ad-hoc experiments.)

There is also a sense in which it's fruitful to run a duplicated rig outside the reporting flow so we can ssh into it and poke around. I'd like to see _less_ of that over time but it'll be the main method of working at first. Implicit in the above flow is an expertise in both vector internals -- so you know where to look for issues -- and in the structure of the test(s) itself. I think this is expertise we can develop in the team.

Reply:
> This actually indicates a need to run ad-hoc experiments in this rig in a way I didn't fully appreciate

This is an interesting piece. I have almost always done ad hoc experimentation locally, just due to the dramatically faster iteration speed, but having some way to close the gap between the two would certainly be useful. There have certainly been cases where my local build and testing results were directionally different from what showed up with a full CI release build and the test harness, though that shouldn't be the norm.

Reply:
I believe that working locally will be the _norm_ but I want to leave the possibility open for hard to replicate setups.

Reply:
Hah, it's likely that our personal machines are going to be dramatically faster than a lot of the VMs our users run Vector on, as well.